src/java.base/share/classes/jdk/internal/lang/StableArray.java

```diff
@@ -0,0 +1,176 @@
+package jdk.internal.lang;
+
+import jdk.internal.lang.stable.ShapeImpl;
+import jdk.internal.lang.stable.StableArrayOne;
+import jdk.internal.lang.stable.StableArrayZero;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Objects;
+
+/**
+ * An atomic, thread-safe, stable array holder for which components can be set at most once.
+ * <p>
+ * Stable arrays are eligible for certain optimizations by the JVM. For multidimensional
+ * arrays, they are also subject to array flattening.
+ * <p>
+ * The total number of components ({@linkplain Shape#size()})in stable arrays can not exceed
+ * about 2<sup>31</sup>.
+ *
+ * Except for a StableArray's component values, all method parameters must be
+ * <em>non-null</em> {@link NullPointerException} will be thrown.
+ *
+ * @param <V> type of StableValue that this stable array holds
+ * @since 23
+ */
+public sealed interface StableArray<V>
```

Comment:
Do we have a use case for StableArray beyond those of StableList?

Reply:
I am trying to model multi-dimensional arrays that also provide flattening. Let's see if it becomes useful.

Reply:
I think this StableArray can be used as an explicit field type to block reflection modifications and enforce constant-folding; the List interface cannot. Though in long term I still believe strict final fields are better.